### PR TITLE
Fix particle boundary condition with mutiple plugins and compositions.

### PR DESCRIPTION
**Pull request by MFraters**
@gassmoeller and I found an issue where an assert with the particle boundary conditions was triggered. The assert was a bit puzzling, but looking at the code we found that it didn't work for our use-case, because the incorrect composition was passed in. This pull request fixes that.

### CODE REVIEW

**Comment by gassmoeller** [comment]
I am certain this is a correct solution, but I am struggling to reproduce the problem we were solving at the time. Do you recall the exact problematic setup? I was trying to reproduce the problem with the attached simple .prm file, but it works (even though I think it shouldnt). We need a test case and I would like to simplify the code before merging this.

[particle_property_initial_composition_boundary_2.prm.txt](https://github.com/geodynamics/aspect/files/13779279/particle_property_initial_composition_boundary_2.prm.txt)

**Comment by gassmoeller** [comment]
@MFraters sorry for hijacking this PR, but I encountered this bug again, and this time I could create a reproducible test case. I uploaded everything to this PR, because it is essentially still the same fix that we discussed together a while ago.

@jdannberg: This PR fixes the most severe bugs of the ones we discussed today. There is still more to improve about the handling of the particle property grain size close to boundaries, but this PR contains two clear bug fixes: 1. The grain size material model currently does not work nicely with multiple compositions, because the plasticity parameters are not correctly parsed. 2. The grain size particle property can be incorrectly initialized from other compositional fields when new particles are generated later in the model run. This results in some particles having absurd grain size values (like 3000, if they happen to be initialized from an entropy field). The attached test has this problem on the main branch (new particles are generated with a grain size of 1000, the value for the boundary composition of the other field), but not on this fixed branch.

Could either of you take a look and let me know if this is ok?

File: source/material_model/grain_size.cc [hunk 1]
```diff
@@ -1074,8 +1074,8 @@ namespace aspect
           use_adiabatic_pressure_for_yielding = prm.get_bool ("Use adiabatic pressure for yield stress");
           drucker_prager_plasticity.initialize_simulator (this->get_simulator());
 
-          std::vector<unsigned int> n_phases = {n_phase_transitions+1};
-          drucker_prager_plasticity.parse_parameters(prm, std::make_unique<std::vector<unsigned int>> (n_phases));
+          const auto n_phase_transitions_for_each_chemical_composition = std::make_unique<std::vector<unsigned int>> (phase_function->n_phase_transitions_for_each_chemical_composition());
+          drucker_prager_plasticity.parse_parameters(prm, n_phase_transitions_for_each_chemical_composition);
```
Comment (jdannberg) [inline source/material_model/grain_size.cc:1078]: I think you fixed this in a different way in #6269?

File: source/particle/property/interface.cc [hunk 2]
```diff
@@ -581,15 +581,23 @@ namespace aspect
...
+                      for (unsigned int property_component = 0; property_component < property_information.get_components_by_plugin_index(property_index); ++property_component)
                         {
-                          const double composition = manager.boundary_composition(boundary_id,particle_location,c);
-                          particle_properties.push_back(composition);
+                          unsigned int composition_index = numbers::invalid_unsigned_int;
+                          const std::string name = property_information.get_field_name_by_index(property_index);
+
+                          for (const std::pair<unsigned int,std::pair<std::string,unsigned int>> item : this->get_parameters().mapped_particle_properties)
+                            {
+                              if (item.second.first == name && item.second.second == property_component)
```
Comment (jdannberg) [inline source/particle/property/interface.cc:593]: Can you add a comment what `item.second.first` and `item.second.second` are?

File: source/particle/property/interface.cc [hunk 3]
```diff
@@ -581,15 +581,23 @@ namespace aspect
...
+                                composition_index = item.first;
+                            }
+                          Assert(composition_index != numbers::invalid_unsigned_int,
+                                 ExcMessage("Could not find composition index " + std::to_string(property_index)
+                                            + " with name " + name + " in the list of compositions."));
+
+                          particle_properties.push_back(manager.boundary_composition(boundary_id,particle_location,composition_index));
```
Comment (jdannberg) [inline source/particle/property/interface.cc:600]: Can you add spaces after the commas?

File: source/particle/property/interface.cc [hunk 4]
```diff
@@ -581,15 +581,23 @@ namespace aspect
...
+                          const std::string name = property_information.get_field_name_by_index(property_index);
+
+                          for (const std::pair<unsigned int,std::pair<std::string,unsigned int>> item : this->get_parameters().mapped_particle_properties)
+                            {
+                              if (item.second.first == name && item.second.second == property_component)
+                                composition_index = item.first;
+                            }
+                          Assert(composition_index != numbers::invalid_unsigned_int,
+                                 ExcMessage("Could not find composition index " + std::to_string(property_index)
+                                            + " with name " + name + " in the list of compositions."));
```
Comment (jdannberg) [inline source/particle/property/interface.cc:598]: Can you check this in the beginning when we create the particles properties (and then you can also do it in release mode).

**Review by jdannberg — COMMENTED** [review]
Thank you for fixing this! The changes in the code make sense to me, but I have a few comments on smaller things I thought could be improved.

**Comment by jdannberg** [comment]
Oh and the `particle_property_initial_composition_boundary` and `viscoelastic_bending_beam_particles` tests are now failing with your new Assert (which is probably exactly due to what you mentioned, that these have passive particles and try to interpolate from the boundary conditions anyway).

**Comment by gassmoeller** [comment]
@jdannberg I found a simpler and more flexible solution. Instead of trying to guess which particle property plugins belong to which compositional field, we should let the particle plugins tell us. They need to know this anyway, as they are the ones requesting to use the boundary condition of the compositional fields by specifying in `late_initialization_mode()` that they want to interpolate from the composition boundary condition.

I think I addressed all your comments, I just havent thought of a solution for what to do, when we have a temperature particle property that wants to interpolate from the temperature boundary condition. Maybe I need to let the particle properties return an `advection_field_index` (composition_index+1)? or let them return an `AdvectionField` class that uniquely defines a field (but that is inside simulator.h, which I dont want to include here). I need to think about this a bit.

**Comment by gassmoeller** [comment]
@jdannberg I replaced the use of compositional indices with the use of `AdvectionField`. This will make this more compatible in case we want to include temperature on particles in the future. Can you take a look?

File: include/aspect/particle/property/composition_reaction.h [hunk 5]
```diff
@@ -78,6 +78,17 @@ namespace aspect
           InitializationModeForLateParticles
           late_initialization_mode () const override;
 
+          /**
+           * A function that returns which advection field to be used
```
Comment (jdannberg) [inline include/aspect/particle/property/composition_reaction.h:82]:
```suggestion
           * A function that returns the advection field to be used
```
Or which advection field **is** to be used

Comment (jdannberg) [reply include/aspect/particle/property/composition_reaction.h:82]: And the same for all the places below.

**Review by jdannberg — APPROVED** [review]
This looks good to me!
I only suggested a few changes with regards to grammar.

**Comment by gassmoeller** [comment]
@jdannberg Thank you for the review! I addressed your comments and squashed the commits. This should be ready to merge.

**Comment by gassmoeller** [comment]
@jdannberg could you take another look?
